### ISSUE TITLE
Bugs in Merging Windowing process

### ISSUE THREAD

**Issue by vanekjar** [seznam/euphoria#79]
When I was going through the code handling window merging I realized there must clearly a few bugs. Although it is not covered by unit test. I added more complex unit test in branch [79/MergingWindowingBug](https://github.com/seznam/euphoria/compare/79/MergingWindowingBug) that is currently failing.

The test uses merging windowing that is uncommon in two attributes:

 * Windows are always triggered only by `onMerge` trigger.
 * Newly created window is never result of the merge, it just triggers the merge on other windows.

Running this test will fail on both executors:

 * In Flink `AbstractWindowOperator` flushes the window `onMerge` even before the window state is merged to the final result. That leads to incorrect result. 
https://github.com/seznam/euphoria/blob/master/euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/windowing/AbstractWindowOperator.java#L173

 * In Spark `GroupReducer` ignores the trigger result of `onMerge` operation in case the newly created window is not part of the merge result.
https://github.com/seznam/euphoria/blob/master/euphoria-core/src/main/java/cz/seznam/euphoria/core/executor/greduce/GroupReducer.java#L218

Please have a look and let me know if my observation is right.

**Comment by xitep** [comment]
Thanks! I went through both places you pointed out, and, yes, I think you are correct. In a broader sense I think, your analysis suggests that we better rethink the possibility for triggers to signal "flush/purge" upon `onMerge`:

* _fact:_ Signaling "flush/purge" always targets the "merge target" window since the "merge source" windows are being implicitly purged as part of being merged.
* _fact:_ A newly arrived element, which triggers the "merge" functionality, is not yet processed and hence not reflected in any state. (this has good reasons since it reduces the times the merging logic is executed.)
  * _problem:_ So, flushing upon "merge" has the semantics of "merging" _just before_ the element (think "trying to merge at random points in time just to see what's possible"). However, this is not fully correct, since the newly arrived element potentially created a new, empty window which is then subject to merge. If that window happened to be the merge target, it might get flushed/purged. Later, that window will incorporate the new element and receive a flush/purge (again) due to an `onElement` trigger. This is a conceptual problem.

As far as I can see, we don't need `onMerge` to signal "flushing/merging" and can avoid the undefined situation by leaving triggering of flush/purge to `onElement` and/or `onTimer`. (Btw: I just noticed that flink 1.2.0 has [some similar reasoning](https://issues.apache.org/jira/browse/FLINK-4993) and also dropped the "trigger-result" from `onMerge`.)
